// wumpcord/build/events/presence/PresenceUpdateEvent.js
"use strict";
/**
 * Copyright (c) 2020-2021 August, Ice
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const models_1 = require("../../models");
const Event_1 = __importDefault(require("../Event"));
class PresenceUpdateEvent extends Event_1.default {
    get new() {
        return this.$refs.presence;
    }
    get old() {
        return this.$refs.old;
    }
    get user() {
        return this.$refs.user;
    }
    get guild() {
        return this.$refs.guild;
    }
    process() {
        const guild = this.data.guild_id !== undefined
            ? this.client.guilds.get(this.data.guild_id)
            : null;
        const old = guild?.presences.get(this.data.user.id) ?? null;
        const user = this.client.users.get(this.data.user.id) || { id: this.data.user.id };
        if (user.id !== this.data.user.id)
            return;
        this.$refs = {
            presence: new models_1.Presence(this.client, this.data),
            guild,
            user,
            old
        };
    }
}
exports.default = PresenceUpdateEvent;
