// wumpcord/build/events/message/index.js
"use strict";
/**
 * Copyright (c) 2020-2021 August, Ice
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageCreateEvent = exports.MessageUpdateEvent = exports.MessageDeleteEvent = exports.MessageDeleteBulkEvent = exports.MessageReactionAddEvent = exports.MessageReactionRemoveEvent = exports.MessageReactionRemoveAllEvent = exports.MessageReactionRemoveEmojiEvent = void 0;
var MessageReactionRemoveEmojiEvent_1 = require("./reaction/MessageReactionRemoveEmojiEvent");
Object.defineProperty(exports, "MessageReactionRemoveEmojiEvent", { enumerable: true, get: function () { return __importDefault(MessageReactionRemoveEmojiEvent_1).default; } });
var MessageReactionRemoveAllEvent_1 = require("./reaction/MessageReactionRemoveAllEvent");
Object.defineProperty(exports, "MessageReactionRemoveAllEvent", { enumerable: true, get: function () { return __importDefault(MessageReactionRemoveAllEvent_1).default; } });
var MessageReactionRemoveEvent_1 = require("./reaction/MessageReactionRemoveEvent");
Object.defineProperty(exports, "MessageReactionRemoveEvent", { enumerable: true, get: function () { return __importDefault(MessageReactionRemoveEvent_1).default; } });
var MessageReactionAddEvent_1 = require("./reaction/MessageReactionAddEvent");
Object.defineProperty(exports, "MessageReactionAddEvent", { enumerable: true, get: function () { return __importDefault(MessageReactionAddEvent_1).default; } });
var MessageDeleteBulkEvent_1 = require("./MessageDeleteBulkEvent");
Object.defineProperty(exports, "MessageDeleteBulkEvent", { enumerable: true, get: function () { return __importDefault(MessageDeleteBulkEvent_1).default; } });
var MessageDeleteEvent_1 = require("./MessageDeleteEvent");
Object.defineProperty(exports, "MessageDeleteEvent", { enumerable: true, get: function () { return __importDefault(MessageDeleteEvent_1).default; } });
var MessageUpdateEvent_1 = require("./MessageUpdateEvent");
Object.defineProperty(exports, "MessageUpdateEvent", { enumerable: true, get: function () { return __importDefault(MessageUpdateEvent_1).default; } });
var MessageCreateEvent_1 = require("./MessageCreateEvent");
Object.defineProperty(exports, "MessageCreateEvent", { enumerable: true, get: function () { return __importDefault(MessageCreateEvent_1).default; } });
