// wumpcord/build/models/audits/AuditLogEntry.js
"use strict";
/**
 * Copyright (c) 2020-2021 August, Ice
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Constants_1 = require("../../Constants");
const Base_1 = __importDefault(require("../Base"));
class AuditLogEntry extends Base_1.default {
    /**
     * Creates a new [AuditLogEntry] instance
     * @param client The WebSocket client attached
     * @param data The data from Discord
     */
    constructor(client, data) {
        super(data.id);
        this.client = client;
        this.patch(data);
    }
    patch(data) {
        this.actionType = data.action_type;
        this.changes = data.changes || [];
        this.reason = data.reason;
        this.user = data.user_id;
        // https://discord.com/developers/docs/resources/audit-log#audit-log-entry-object-optional-audit-entry-info
        if (data.options !== undefined) {
            if (this.actionType === Constants_1.AuditLogAction.MemberPrune) {
                this.inactiveMembers = Number(data.options.delete_member_days);
                this.membersRemoved = Number(data.options.members_removed);
            }
            if ([
                Constants_1.AuditLogAction.MemberMove,
                Constants_1.AuditLogAction.MessagePin,
                Constants_1.AuditLogAction.MessageUnpin,
                Constants_1.AuditLogAction.MessageDelete
            ].includes(this.actionType)) {
                this.channelID = data.options.channel_id;
                this.channel = undefined;
            }
            if ([Constants_1.AuditLogAction.MessageUnpin, Constants_1.AuditLogAction.MessagePin].includes(this.actionType)) {
                this.messageID = data.options.message_id;
                this.message = undefined;
            }
            if ([
                Constants_1.AuditLogAction.MessageDelete,
                Constants_1.AuditLogAction.MessageBulkDelete,
                Constants_1.AuditLogAction.MemberDisconnect,
                Constants_1.AuditLogAction.MemberMove
            ].includes(this.actionType)) {
                this.targetCount = Number(data.options.count);
            }
            if ([
                Constants_1.AuditLogAction.ChannelOverwriteCreate,
                Constants_1.AuditLogAction.ChannelOverwriteDelete,
                Constants_1.AuditLogAction.ChannelOverwriteUpdate
            ].includes(this.actionType)) {
                this.type = data.options.type === '0' ? 'role' : 'member';
                if (data.options.type === '0')
                    this.role = data.options.role_name;
            }
        }
    }
}
exports.default = AuditLogEntry;
