// wumpcord/build/events/guild/index.js
"use strict";
/**
 * Copyright (c) 2020-2021 August, Ice
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuildDeleteEvent = exports.GuildCreateEvent = exports.GuildUpdateEvent = exports.GuildEmojisUpdateEvent = exports.GuildIntegrationsUpdateEvent = exports.GuildRoleCreateEvent = exports.GuildRoleUpdateEvent = exports.GuildRoleDeleteEvent = exports.GuildMemberAddEvent = exports.GuildMemberChunkEvent = exports.GuildMemberUpdateEvent = exports.GuildMemberRemoveEvent = exports.GuildBanAddEvent = exports.GuildBanRemoveEvent = void 0;
// Guild Bans
var GuildBanRemoveEvent_1 = require("./bans/GuildBanRemoveEvent");
Object.defineProperty(exports, "GuildBanRemoveEvent", { enumerable: true, get: function () { return __importDefault(GuildBanRemoveEvent_1).default; } });
var GuildBanAddEvent_1 = require("./bans/GuildBanAddEvent");
Object.defineProperty(exports, "GuildBanAddEvent", { enumerable: true, get: function () { return __importDefault(GuildBanAddEvent_1).default; } });
// Guild Member
var GuildMemberRemoveEvent_1 = require("./member/GuildMemberRemoveEvent");
Object.defineProperty(exports, "GuildMemberRemoveEvent", { enumerable: true, get: function () { return __importDefault(GuildMemberRemoveEvent_1).default; } });
var GuildMemberUpdateEvent_1 = require("./member/GuildMemberUpdateEvent");
Object.defineProperty(exports, "GuildMemberUpdateEvent", { enumerable: true, get: function () { return __importDefault(GuildMemberUpdateEvent_1).default; } });
var GuildMemberChunkEvent_1 = require("./member/GuildMemberChunkEvent");
Object.defineProperty(exports, "GuildMemberChunkEvent", { enumerable: true, get: function () { return __importDefault(GuildMemberChunkEvent_1).default; } });
var GuildMemberAddEvent_1 = require("./member/GuildMemberAddEvent");
Object.defineProperty(exports, "GuildMemberAddEvent", { enumerable: true, get: function () { return __importDefault(GuildMemberAddEvent_1).default; } });
// Guild Roles
var GuildRoleDeleteEvent_1 = require("./roles/GuildRoleDeleteEvent");
Object.defineProperty(exports, "GuildRoleDeleteEvent", { enumerable: true, get: function () { return __importDefault(GuildRoleDeleteEvent_1).default; } });
var GuildRoleUpdateEvent_1 = require("./roles/GuildRoleUpdateEvent");
Object.defineProperty(exports, "GuildRoleUpdateEvent", { enumerable: true, get: function () { return __importDefault(GuildRoleUpdateEvent_1).default; } });
var GuildRoleCreateEvent_1 = require("./roles/GuildRoleCreateEvent");
Object.defineProperty(exports, "GuildRoleCreateEvent", { enumerable: true, get: function () { return __importDefault(GuildRoleCreateEvent_1).default; } });
// Guild
var GuildIntegrationsUpdateEvent_1 = require("./GuildIntegrationsUpdateEvent");
Object.defineProperty(exports, "GuildIntegrationsUpdateEvent", { enumerable: true, get: function () { return __importDefault(GuildIntegrationsUpdateEvent_1).default; } });
var GuildEmojisUpdateEvent_1 = require("./GuildEmojisUpdateEvent");
Object.defineProperty(exports, "GuildEmojisUpdateEvent", { enumerable: true, get: function () { return __importDefault(GuildEmojisUpdateEvent_1).default; } });
var GuildUpdateEvent_1 = require("./GuildUpdateEvent");
Object.defineProperty(exports, "GuildUpdateEvent", { enumerable: true, get: function () { return __importDefault(GuildUpdateEvent_1).default; } });
var GuildCreateEvent_1 = require("./GuildCreateEvent");
Object.defineProperty(exports, "GuildCreateEvent", { enumerable: true, get: function () { return __importDefault(GuildCreateEvent_1).default; } });
var GuildDeleteEvent_1 = require("./GuildDeleteEvent");
Object.defineProperty(exports, "GuildDeleteEvent", { enumerable: true, get: function () { return __importDefault(GuildDeleteEvent_1).default; } });
