// discord-api-types/v8/payloads/interactions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIInteractionResponseType = exports.InteractionType = exports.ApplicationCommandOptionType = void 0;
/**
 * https://discord.com/developers/docs/interactions/slash-commands#applicationcommandoptiontype
 */
var ApplicationCommandOptionType;
(function (ApplicationCommandOptionType) {
    ApplicationCommandOptionType[ApplicationCommandOptionType["SUB_COMMAND"] = 1] = "SUB_COMMAND";
    ApplicationCommandOptionType[ApplicationCommandOptionType["SUB_COMMAND_GROUP"] = 2] = "SUB_COMMAND_GROUP";
    ApplicationCommandOptionType[ApplicationCommandOptionType["STRING"] = 3] = "STRING";
    ApplicationCommandOptionType[ApplicationCommandOptionType["INTEGER"] = 4] = "INTEGER";
    ApplicationCommandOptionType[ApplicationCommandOptionType["BOOLEAN"] = 5] = "BOOLEAN";
    ApplicationCommandOptionType[ApplicationCommandOptionType["USER"] = 6] = "USER";
    ApplicationCommandOptionType[ApplicationCommandOptionType["CHANNEL"] = 7] = "CHANNEL";
    ApplicationCommandOptionType[ApplicationCommandOptionType["ROLE"] = 8] = "ROLE";
})(ApplicationCommandOptionType = exports.ApplicationCommandOptionType || (exports.ApplicationCommandOptionType = {}));
/**
 * https://discord.com/developers/docs/interactions/slash-commands#interaction-interactiontype
 */
var InteractionType;
(function (InteractionType) {
    InteractionType[InteractionType["Ping"] = 1] = "Ping";
    InteractionType[InteractionType["ApplicationCommand"] = 2] = "ApplicationCommand";
})(InteractionType = exports.InteractionType || (exports.InteractionType = {}));
/**
 * https://discord.com/developers/docs/interactions/slash-commands#interaction-interactionresponsetype
 */
var APIInteractionResponseType;
(function (APIInteractionResponseType) {
    APIInteractionResponseType[APIInteractionResponseType["Pong"] = 1] = "Pong";
    APIInteractionResponseType[APIInteractionResponseType["Acknowledge"] = 2] = "Acknowledge";
    APIInteractionResponseType[APIInteractionResponseType["ChannelMessage"] = 3] = "ChannelMessage";
    APIInteractionResponseType[APIInteractionResponseType["ChannelMessageWithSource"] = 4] = "ChannelMessageWithSource";
    APIInteractionResponseType[APIInteractionResponseType["AcknowledgeWithSource"] = 5] = "AcknowledgeWithSource";
})(APIInteractionResponseType = exports.APIInteractionResponseType || (exports.APIInteractionResponseType = {}));
