// @augu/orchid/build/middleware/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.blobs = exports.forms = exports.logging = exports.streams = exports.compress = exports.CycleType = void 0;
const EnableCompressedData_1 = __importDefault(require("./EnableCompressedData"));
exports.compress = EnableCompressedData_1.default;
const EnableStreams_1 = __importDefault(require("./EnableStreams"));
exports.streams = EnableStreams_1.default;
const Logging_1 = __importDefault(require("./Logging"));
exports.logging = Logging_1.default;
const FormData_1 = __importDefault(require("./FormData"));
exports.forms = FormData_1.default;
const Blob_1 = __importDefault(require("./Blob"));
exports.blobs = Blob_1.default;
var CycleType;
(function (CycleType) {
    CycleType["Execute"] = "execute";
    CycleType["Done"] = "done";
    CycleType["None"] = "none";
})(CycleType = exports.CycleType || (exports.CycleType = {}));
