// wumpcord/build/Constants.js
"use strict";
/**
 * Copyright (c) 2020-2021 August, Ice
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StickerType = exports.ShardStatus = exports.ActivityFlags = exports.GatewayEvents = exports.GatewayIntents = exports.UserFlags = exports.ChannelTypesObject = exports.ChannelTypes = exports.MessageTypes = exports.MessageFlags = exports.AuditLogAction = exports.Permissions = exports.OPCodes = exports.ActivityStatus = exports.WebhookTypes = exports.RestUrl = exports.CDNUrl = exports.UserAgent = exports.ImageFormats = exports.RestVersion = exports.GatewayVersion = exports.UnrecoverableCodes = exports.StrategyTypes = void 0;
const { version } = require('../package.json');
exports.StrategyTypes = ['etf', 'json'];
exports.UnrecoverableCodes = [
    1005,
    4004,
    4010,
    4011,
    4013,
    4014 // disallowed intents
];
exports.GatewayVersion = 8;
exports.RestVersion = 8;
exports.ImageFormats = [
    'png',
    'jpg',
    'gif',
    'webp',
    'jpeg'
];
exports.UserAgent = `DiscordBot (https://github.com/auguwu/Wumpcord, v${version})`;
exports.CDNUrl = 'https://cdn.discordapp.com';
exports.RestUrl = 'https://discord.com/api';
// The first type index at 1, so it's gonna be null
exports.WebhookTypes = [null, 'Incoming', 'Channel Following'];
var ActivityStatus;
(function (ActivityStatus) {
    ActivityStatus[ActivityStatus["Playing"] = 0] = "Playing";
    ActivityStatus[ActivityStatus["Streaming"] = 1] = "Streaming";
    ActivityStatus[ActivityStatus["Listening"] = 2] = "Listening";
    // 3 is hidden (Watching)
    ActivityStatus[ActivityStatus["Custom"] = 4] = "Custom";
    ActivityStatus[ActivityStatus["Competing"] = 5] = "Competing";
})(ActivityStatus = exports.ActivityStatus || (exports.ActivityStatus = {}));
var OPCodes;
(function (OPCodes) {
    OPCodes[OPCodes["Event"] = 0] = "Event";
    OPCodes[OPCodes["Heartbeat"] = 1] = "Heartbeat";
    OPCodes[OPCodes["Identify"] = 2] = "Identify";
    OPCodes[OPCodes["StatusUpdate"] = 3] = "StatusUpdate";
    OPCodes[OPCodes["VoiceServerUpdate"] = 4] = "VoiceServerUpdate";
    // 5 is a mystery
    OPCodes[OPCodes["Resume"] = 6] = "Resume";
    OPCodes[OPCodes["Reconnect"] = 7] = "Reconnect";
    OPCodes[OPCodes["GetGuildMembers"] = 8] = "GetGuildMembers";
    OPCodes[OPCodes["InvalidSession"] = 9] = "InvalidSession";
    OPCodes[OPCodes["Hello"] = 10] = "Hello";
    OPCodes[OPCodes["HeartbeatAck"] = 11] = "HeartbeatAck";
    OPCodes[OPCodes["SyncGuild"] = 12] = "SyncGuild";
    OPCodes[OPCodes["SyncCall"] = 13] = "SyncCall";
})(OPCodes = exports.OPCodes || (exports.OPCodes = {}));
var Permissions;
(function (Permissions) {
    Permissions[Permissions["createInstantInvite"] = 1] = "createInstantInvite";
    Permissions[Permissions["kickMembers"] = 2] = "kickMembers";
    Permissions[Permissions["banMembers"] = 4] = "banMembers";
    Permissions[Permissions["administrator"] = 8] = "administrator";
    Permissions[Permissions["manageChannels"] = 16] = "manageChannels";
    Permissions[Permissions["manageGuild"] = 32] = "manageGuild";
    Permissions[Permissions["addReactions"] = 64] = "addReactions";
    Permissions[Permissions["viewAuditLogs"] = 128] = "viewAuditLogs";
    Permissions[Permissions["voicePrioritySpeaker"] = 256] = "voicePrioritySpeaker";
    Permissions[Permissions["stream"] = 512] = "stream";
    Permissions[Permissions["readMessages"] = 1024] = "readMessages";
    Permissions[Permissions["sendMessages"] = 2048] = "sendMessages";
    Permissions[Permissions["sendTTSMessages"] = 4096] = "sendTTSMessages";
    Permissions[Permissions["manageMessages"] = 8192] = "manageMessages";
    Permissions[Permissions["embedLinks"] = 16384] = "embedLinks";
    Permissions[Permissions["attachFiles"] = 32768] = "attachFiles";
    Permissions[Permissions["readMessageHistory"] = 65536] = "readMessageHistory";
    Permissions[Permissions["mentionEveryone"] = 131072] = "mentionEveryone";
    Permissions[Permissions["externalEmojis"] = 262144] = "externalEmojis";
    Permissions[Permissions["viewGuildInsights"] = 524288] = "viewGuildInsights";
    Permissions[Permissions["voiceConnect"] = 1048576] = "voiceConnect";
    Permissions[Permissions["voiceSpeak"] = 2097152] = "voiceSpeak";
    Permissions[Permissions["voiceMuteMembers"] = 4194304] = "voiceMuteMembers";
    Permissions[Permissions["voiceDeafenMembers"] = 8388608] = "voiceDeafenMembers";
    Permissions[Permissions["voiceMoveMembers"] = 16777216] = "voiceMoveMembers";
    Permissions[Permissions["voiceUseVAD"] = 33554432] = "voiceUseVAD";
    Permissions[Permissions["changeNickname"] = 67108864] = "changeNickname";
    Permissions[Permissions["manageNicknames"] = 134217728] = "manageNicknames";
    Permissions[Permissions["manageRoles"] = 268435456] = "manageRoles";
    Permissions[Permissions["manageWebhooks"] = 536870912] = "manageWebhooks";
    Permissions[Permissions["manageEmojis"] = 1073741824] = "manageEmojis";
    Permissions[Permissions["all"] = 2147483647] = "all";
    Permissions[Permissions["allGuild"] = 2080899263] = "allGuild";
    Permissions[Permissions["allText"] = 805829713] = "allText";
    Permissions[Permissions["allVoice"] = 871367441] = "allVoice";
})(Permissions = exports.Permissions || (exports.Permissions = {}));
var AuditLogAction;
(function (AuditLogAction) {
    AuditLogAction[AuditLogAction["GuildUpdate"] = 1] = "GuildUpdate";
    AuditLogAction[AuditLogAction["ChannelCreate"] = 10] = "ChannelCreate";
    AuditLogAction[AuditLogAction["ChannelUpdate"] = 11] = "ChannelUpdate";
    AuditLogAction[AuditLogAction["ChannelDelete"] = 12] = "ChannelDelete";
    AuditLogAction[AuditLogAction["ChannelOverwriteCreate"] = 13] = "ChannelOverwriteCreate";
    AuditLogAction[AuditLogAction["ChannelOverwriteUpdate"] = 14] = "ChannelOverwriteUpdate";
    AuditLogAction[AuditLogAction["ChannelOverwriteDelete"] = 15] = "ChannelOverwriteDelete";
    AuditLogAction[AuditLogAction["MemberKick"] = 20] = "MemberKick";
    AuditLogAction[AuditLogAction["MemberPrune"] = 21] = "MemberPrune";
    AuditLogAction[AuditLogAction["MemberBanAdd"] = 22] = "MemberBanAdd";
    AuditLogAction[AuditLogAction["MemberBanRemove"] = 23] = "MemberBanRemove";
    AuditLogAction[AuditLogAction["MemberUpdate"] = 24] = "MemberUpdate";
    AuditLogAction[AuditLogAction["MemberRoleUpdate"] = 25] = "MemberRoleUpdate";
    AuditLogAction[AuditLogAction["MemberMove"] = 26] = "MemberMove";
    AuditLogAction[AuditLogAction["MemberDisconnect"] = 27] = "MemberDisconnect";
    AuditLogAction[AuditLogAction["BotAdd"] = 28] = "BotAdd";
    AuditLogAction[AuditLogAction["RoleCreate"] = 30] = "RoleCreate";
    AuditLogAction[AuditLogAction["RoleUpdate"] = 31] = "RoleUpdate";
    AuditLogAction[AuditLogAction["RoleDelete"] = 32] = "RoleDelete";
    AuditLogAction[AuditLogAction["InviteCreate"] = 40] = "InviteCreate";
    AuditLogAction[AuditLogAction["InviteUpdate"] = 41] = "InviteUpdate";
    AuditLogAction[AuditLogAction["InviteDelete"] = 42] = "InviteDelete";
    AuditLogAction[AuditLogAction["WebhookCreate"] = 50] = "WebhookCreate";
    AuditLogAction[AuditLogAction["WebhookUpdate"] = 51] = "WebhookUpdate";
    AuditLogAction[AuditLogAction["WebhookDelete"] = 52] = "WebhookDelete";
    AuditLogAction[AuditLogAction["EmojiCreate"] = 60] = "EmojiCreate";
    AuditLogAction[AuditLogAction["EmojiUpdate"] = 61] = "EmojiUpdate";
    AuditLogAction[AuditLogAction["EmojiDelete"] = 62] = "EmojiDelete";
    AuditLogAction[AuditLogAction["MessageDelete"] = 72] = "MessageDelete";
    AuditLogAction[AuditLogAction["MessageBulkDelete"] = 73] = "MessageBulkDelete";
    AuditLogAction[AuditLogAction["MessagePin"] = 74] = "MessagePin";
    AuditLogAction[AuditLogAction["MessageUnpin"] = 75] = "MessageUnpin";
    AuditLogAction[AuditLogAction["IntegrationCreate"] = 80] = "IntegrationCreate";
    AuditLogAction[AuditLogAction["IntegrationUpdate"] = 81] = "IntegrationUpdate";
    AuditLogAction[AuditLogAction["IntegrationDelete"] = 82] = "IntegrationDelete";
})(AuditLogAction = exports.AuditLogAction || (exports.AuditLogAction = {}));
var MessageFlags;
(function (MessageFlags) {
    MessageFlags[MessageFlags["Crossposted"] = 1] = "Crossposted";
    MessageFlags[MessageFlags["IsCrosspost"] = 2] = "IsCrosspost";
    MessageFlags[MessageFlags["SupressEmbed"] = 4] = "SupressEmbed";
    MessageFlags[MessageFlags["SourceMessageDeleted"] = 8] = "SourceMessageDeleted";
    MessageFlags[MessageFlags["Urgent"] = 16] = "Urgent";
})(MessageFlags = exports.MessageFlags || (exports.MessageFlags = {}));
var MessageTypes;
(function (MessageTypes) {
    MessageTypes[MessageTypes["Default"] = 0] = "Default";
    MessageTypes[MessageTypes["RecipientAdd"] = 1] = "RecipientAdd";
    MessageTypes[MessageTypes["RecipientRemove"] = 2] = "RecipientRemove";
    MessageTypes[MessageTypes["Call"] = 3] = "Call";
    MessageTypes[MessageTypes["ChannelNameChange"] = 4] = "ChannelNameChange";
    MessageTypes[MessageTypes["ChannelIconChange"] = 5] = "ChannelIconChange";
    MessageTypes[MessageTypes["ChannelPinnedMessage"] = 6] = "ChannelPinnedMessage";
    MessageTypes[MessageTypes["GuildMemberJoin"] = 7] = "GuildMemberJoin";
    MessageTypes[MessageTypes["UserPremiumGuildSubscription"] = 8] = "UserPremiumGuildSubscription";
    MessageTypes[MessageTypes["UserPremiumGuildSubscriptionTier1"] = 9] = "UserPremiumGuildSubscriptionTier1";
    MessageTypes[MessageTypes["UserPremiumGuildSubscriptionTier2"] = 10] = "UserPremiumGuildSubscriptionTier2";
    MessageTypes[MessageTypes["UserPremiumGuildSubscriptionTier3"] = 11] = "UserPremiumGuildSubscriptionTier3";
    MessageTypes[MessageTypes["ChannelFollowAdd"] = 12] = "ChannelFollowAdd";
    MessageTypes[MessageTypes["GuildDiscoveryDisqualified"] = 15] = "GuildDiscoveryDisqualified";
    MessageTypes[MessageTypes["GuildDiscoveryRequalified"] = 16] = "GuildDiscoveryRequalified";
})(MessageTypes = exports.MessageTypes || (exports.MessageTypes = {}));
var ChannelTypes;
(function (ChannelTypes) {
    ChannelTypes[ChannelTypes["Text"] = 0] = "Text";
    ChannelTypes[ChannelTypes["DM"] = 1] = "DM";
    ChannelTypes[ChannelTypes["Voice"] = 2] = "Voice";
    ChannelTypes[ChannelTypes["Group"] = 3] = "Group";
    ChannelTypes[ChannelTypes["News"] = 4] = "News";
    ChannelTypes[ChannelTypes["Store"] = 5] = "Store";
})(ChannelTypes = exports.ChannelTypes || (exports.ChannelTypes = {}));
exports.ChannelTypesObject = {
    0: 'text',
    1: 'dm',
    2: 'voice',
    3: 'group',
    4: 'category',
    5: 'news',
    6: 'store'
};
var UserFlags;
(function (UserFlags) {
    UserFlags[UserFlags["None"] = 0] = "None";
    UserFlags[UserFlags["Staff"] = 1] = "Staff";
    UserFlags[UserFlags["Partner"] = 2] = "Partner";
    UserFlags[UserFlags["HypesquadEvents"] = 4] = "HypesquadEvents";
    UserFlags[UserFlags["BugHunterLevel1"] = 8] = "BugHunterLevel1";
    UserFlags[UserFlags["Bravery"] = 64] = "Bravery";
    UserFlags[UserFlags["Brillance"] = 128] = "Brillance";
    UserFlags[UserFlags["Balance"] = 256] = "Balance";
    UserFlags[UserFlags["EarlySupporter"] = 512] = "EarlySupporter";
    UserFlags[UserFlags["TeamUser"] = 1024] = "TeamUser";
    UserFlags[UserFlags["System"] = 4096] = "System";
    UserFlags[UserFlags["BugHunterLevel2"] = 16384] = "BugHunterLevel2";
    UserFlags[UserFlags["VerifiedBot"] = 65536] = "VerifiedBot";
    UserFlags[UserFlags["VerifiedBotDev"] = 131072] = "VerifiedBotDev";
})(UserFlags = exports.UserFlags || (exports.UserFlags = {}));
var GatewayIntents;
(function (GatewayIntents) {
    GatewayIntents[GatewayIntents["guilds"] = 1] = "guilds";
    GatewayIntents[GatewayIntents["guildMembers"] = 2] = "guildMembers";
    GatewayIntents[GatewayIntents["guildBans"] = 4] = "guildBans";
    GatewayIntents[GatewayIntents["guildEmojis"] = 8] = "guildEmojis";
    GatewayIntents[GatewayIntents["guildIntegrations"] = 16] = "guildIntegrations";
    GatewayIntents[GatewayIntents["guildWebhooks"] = 32] = "guildWebhooks";
    GatewayIntents[GatewayIntents["guildInvites"] = 64] = "guildInvites";
    GatewayIntents[GatewayIntents["guildVoiceStates"] = 128] = "guildVoiceStates";
    GatewayIntents[GatewayIntents["guildPresences"] = 256] = "guildPresences";
    GatewayIntents[GatewayIntents["guildMessages"] = 512] = "guildMessages";
    GatewayIntents[GatewayIntents["guildMessageReactions"] = 1024] = "guildMessageReactions";
    GatewayIntents[GatewayIntents["guildMessageTyping"] = 2048] = "guildMessageTyping";
    GatewayIntents[GatewayIntents["directMessages"] = 4096] = "directMessages";
    GatewayIntents[GatewayIntents["directMessageReactions"] = 8192] = "directMessageReactions";
    GatewayIntents[GatewayIntents["directMessageTyping"] = 16384] = "directMessageTyping";
})(GatewayIntents = exports.GatewayIntents || (exports.GatewayIntents = {}));
var GatewayEvents;
(function (GatewayEvents) {
    GatewayEvents["Ready"] = "READY";
    GatewayEvents["Resumed"] = "RESUMED";
    GatewayEvents["ChannelCreate"] = "CHANNEL_CREATE";
    GatewayEvents["ChannelUpdate"] = "CHANNEL_UPDATE";
    GatewayEvents["ChannelDelete"] = "CHANNEL_DELETE";
    GatewayEvents["ChannelPinUpdate"] = "CHANNEL_PINS_UPDATE";
    GatewayEvents["GuildCreate"] = "GUILD_CREATE";
    GatewayEvents["GuildUpdate"] = "GUILD_UPDATE";
    GatewayEvents["GuildDelete"] = "GUILD_DELETE";
    GatewayEvents["GuildBanAdd"] = "GUILD_BAN_ADD";
    GatewayEvents["GuildBanRemove"] = "GUILD_BAN_REMOVE";
    GatewayEvents["GuildEmojisUpdate"] = "GUILD_EMOJIS_UPDATE";
    GatewayEvents["GuildIntegrationUpdate"] = "GUILD_INTEGRATIONS_UPDATE";
    GatewayEvents["GuildMemberAdd"] = "GUILD_MEMBER_ADD";
    GatewayEvents["GuildMemberDelete"] = "GUILD_MEMBER_REMOVE";
    GatewayEvents["GuildMemberUpdate"] = "GUILD_MEMBER_UPDATE";
    GatewayEvents["GuildMemberChunk"] = "GUILD_MEMBERS_CHUNK";
    GatewayEvents["GuildRoleCreate"] = "GUILD_ROLE_CREATE";
    GatewayEvents["GuildRoleDelete"] = "GUILD_ROLE_DELETE";
    GatewayEvents["GuildRoleUpdate"] = "GUILD_ROLE_UPDATE";
    GatewayEvents["MessageCreated"] = "MESSAGE_CREATE";
    GatewayEvents["MessageUpdate"] = "MESSAGE_UPDATE";
    GatewayEvents["MessageDelete"] = "MESSAGE_DELETE";
    GatewayEvents["MessageDeleteBulk"] = "MESSAGE_DELETE_BULK";
    GatewayEvents["MessageReactionAdd"] = "MESSAGE_REACTION_ADD";
    GatewayEvents["MessageReactionRemove"] = "MESSAGE_REACTION_REMOVE";
    GatewayEvents["MessageReactionRemoveAll"] = "MESSAGE_REACTION_REMOVE_ALL";
    GatewayEvents["MessageReactionRemoveEmoji"] = "MESSAGE_REACTION_REMOVE_EMOJI";
    GatewayEvents["TypingStart"] = "TYPING_START";
    GatewayEvents["UserUpdate"] = "USER_UPDATE";
    GatewayEvents["VoiceStateUpdate"] = "VOICE_STATE_UPDATE";
    GatewayEvents["VoiceServerUpdate"] = "VOICE_SERVER_UPDATE";
    GatewayEvents["WebhookUpdate"] = "WEBHOOKS_UPDATE";
    GatewayEvents["PresenceUpdate"] = "PRESENCE_UPDATE";
    GatewayEvents["GiftCodeUpdate"] = "GIFT_CODE_UPDATE";
    GatewayEvents["InteractionCreate"] = "INTERACTION_CREATE";
    GatewayEvents["ApplicationCommandUpdate"] = "APPLICATION_COMMAND_UPDATE";
})(GatewayEvents = exports.GatewayEvents || (exports.GatewayEvents = {}));
var ActivityFlags;
(function (ActivityFlags) {
    ActivityFlags[ActivityFlags["Instance"] = 1] = "Instance";
    ActivityFlags[ActivityFlags["Join"] = 2] = "Join";
    ActivityFlags[ActivityFlags["Spectate"] = 4] = "Spectate";
    ActivityFlags[ActivityFlags["JoinRequest"] = 8] = "JoinRequest";
    ActivityFlags[ActivityFlags["Sync"] = 16] = "Sync";
    ActivityFlags[ActivityFlags["Play"] = 32] = "Play";
})(ActivityFlags = exports.ActivityFlags || (exports.ActivityFlags = {}));
var ShardStatus;
(function (ShardStatus) {
    ShardStatus["Connected"] = "connected";
    ShardStatus["Handshaking"] = "handshaking";
    ShardStatus["Nearly"] = "nearly";
    ShardStatus["Dead"] = "dead";
    ShardStatus["WaitingForGuilds"] = "waiting_for_guilds";
})(ShardStatus = exports.ShardStatus || (exports.ShardStatus = {}));
var StickerType;
(function (StickerType) {
    StickerType[StickerType["PNG"] = 0] = "PNG";
    StickerType[StickerType["APNG"] = 1] = "APNG";
    StickerType[StickerType["Lottie"] = 2] = "Lottie";
})(StickerType = exports.StickerType || (exports.StickerType = {}));
