// wumpcord/build/models/channel/GroupChannel.js
"use strict";
/**
 * Copyright (c) 2020-2021 August, Ice
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroupChannel = void 0;
const TextableChannel_1 = __importDefault(require("../inherit/TextableChannel"));
const User_1 = require("../User");
class GroupChannel extends TextableChannel_1.default {
    /**
     * Creates a new [GroupChannel] instance
     * @param client The [WebSocket] client attached
     * @param data The data from Discord
     */
    constructor(client, data) {
        super(client, data);
        this.patch(data);
    }
    patch(data) {
        super.patch(data);
        if (data.last_message_id !== undefined)
            this.lastMessageID = data.last_message_id;
        if (data.last_pin_timestamp !== undefined)
            this.lastPinTimestamp = data.last_pin_timestamp;
        if (data.owner_id !== undefined)
            this.ownerID = data.owner_id;
        if (data.name !== undefined)
            this.name = data.name;
        if (data.icon !== undefined)
            this.icon = data.icon;
        this.recipients = data.recipients.map(data => this.client.users.add(new User_1.User(this.client, data)));
    }
    toString() {
        return `[wumpcord.GroupChannel "${this.name}"]`;
    }
}
exports.GroupChannel = GroupChannel;
