// wumpcord/build/models/channel/NewsChannel.js
"use strict";
/**
 * Copyright (c) 2020-2021 August, Ice
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewsChannel = void 0;
const TextChannel_1 = require("./TextChannel");
const Message_1 = require("../Message");
const util_1 = __importDefault(require("../../util"));
class NewsChannel extends TextChannel_1.TextChannel {
    /**
     * Follows this [NewsChannel] to a different channel
     * @param channelID The channel's ID
     * @returns The result, if it was a success or not
     */
    follow(channelID) {
        return this.client.rest.dispatch({
            endpoint: `/channels/${this.id}/followers`,
            method: 'post',
            data: {
                webhook_channel_id: channelID
            }
        });
    }
    /**
     * Crossposts a message to all the followers who subscribed to this [NewsChannel]
     * @param messageID The message ID to crosspost
     * @param content The content to send out
     * @param options Any additional options, if needed
     * @returns A new [Message] instance indicating it was sent
     */
    crosspost(messageID, content, options) {
        const data = util_1.default.formatMessage(this.client, content, options);
        const file = data.file;
        // delete it so it doesn't bleed when sending
        delete data.file;
        return this.client.rest.dispatch({
            endpoint: `/channels/${this.id}/messages/${messageID}/crosspost`,
            method: 'post',
            file,
            data
        }).then(data => new Message_1.Message(this.client, data));
    }
}
exports.NewsChannel = NewsChannel;
