// wumpcord/build/events/index.js
"use strict";
/**
 * Copyright (c) 2020-2021 August, Ice
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InviteDeleteEvent = exports.InviteCreateEvent = exports.ChannelDeleteEvent = exports.ChannelUpdateEvent = exports.ChannelCreateEvent = exports.ChannelPinsUpdateEvent = void 0;
var ChannelPinsUpdateEvent_1 = require("./channel/ChannelPinsUpdateEvent");
Object.defineProperty(exports, "ChannelPinsUpdateEvent", { enumerable: true, get: function () { return __importDefault(ChannelPinsUpdateEvent_1).default; } });
var ChannelCreateEvent_1 = require("./channel/ChannelCreateEvent");
Object.defineProperty(exports, "ChannelCreateEvent", { enumerable: true, get: function () { return __importDefault(ChannelCreateEvent_1).default; } });
var ChannelUpdateEvent_1 = require("./channel/ChannelUpdateEvent");
Object.defineProperty(exports, "ChannelUpdateEvent", { enumerable: true, get: function () { return __importDefault(ChannelUpdateEvent_1).default; } });
var ChannelDeleteEvent_1 = require("./channel/ChannelDeleteEvent");
Object.defineProperty(exports, "ChannelDeleteEvent", { enumerable: true, get: function () { return __importDefault(ChannelDeleteEvent_1).default; } });
var InviteCreateEvent_1 = require("./invites/InviteCreateEvent");
Object.defineProperty(exports, "InviteCreateEvent", { enumerable: true, get: function () { return __importDefault(InviteCreateEvent_1).default; } });
var InviteDeleteEvent_1 = require("./invites/InviteDeleteEvent");
Object.defineProperty(exports, "InviteDeleteEvent", { enumerable: true, get: function () { return __importDefault(InviteDeleteEvent_1).default; } });
__exportStar(require("./interaction"), exports);
__exportStar(require("./presence"), exports);
__exportStar(require("./message"), exports);
__exportStar(require("./guild"), exports);
__exportStar(require("./voice"), exports);
__exportStar(require("./WebhooksUpdateEvent"), exports);
__exportStar(require("./TypingStartEvent"), exports);
__exportStar(require("./UserUpdateEvent"), exports);
